components/sidebar/sidebar_channel/sidebar_channel.tsx

```diff
@@ -44,31 +45,84 @@ type Props = {
      */
     showUnreadForMsgs: boolean;
 
+    /**
+     * Gets the ref for a given channel id
+     */
+    getChannelRef: (channelId: string) => HTMLDivElement | undefined;
+
     /**
      * Sets the ref for the sidebar channel div element, so that it can be used by parent components
      */
     setChannelRef: (channelId: string, ref: HTMLDivElement) => void;
+
+    /**
+     * If category is collapsed
+     */
+    isCategoryCollapsed: boolean;
+
+    /**
+     * Is the channel the currently focused channel
+     */
+    isCurrentChannel: boolean;
 };
 
 type State = {
-
+    isUnread: boolean;
+    isCollapsed: boolean;
 };
 
 export default class SidebarChannel extends React.PureComponent<Props, State> {
-    /**
-     * Show as unread if you have unread mentions
-     * OR if you have unread messages and the channel can be marked unread by preferences
-     */
-    showChannelAsUnread = () => {
-        return this.props.unreadMentions > 0 || (this.props.unreadMsgs > 0 && this.props.showUnreadForMsgs);
-    };
+    static getDerivedStateFromProps(nextProps: Props) {
+        const isUnread = nextProps.unreadMentions > 0 || (nextProps.unreadMsgs > 0 && nextProps.showUnreadForMsgs);
+        const isCollapsed = nextProps.isCategoryCollapsed && !isUnread && !nextProps.isCurrentChannel;
+
+        return {
+            isUnread,
+            isCollapsed,
+        };
+    }
+
+    componentDidMount() {
+        const channelElement = this.getRef();
+        if (channelElement) {
+            channelElement.addEventListener('transitionend', this.removeAnimation);
```

Comment:
Could this listener be added in the `render` method instead of adding it after the fact?

Reply:
That really doesn't feel right to me, because it's going to be added/removed on every `render()`. Instead this only adds it and removes it once.

Reply:
I don't follow. If you pass an `onTransitionEnd` prop into the div, it'll only be added once when the div is mounted.